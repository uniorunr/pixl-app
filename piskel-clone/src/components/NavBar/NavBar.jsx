import React, { Fragment } from 'react';
import './NavBar.scss';

const NavBar = () => (
  <Fragment>
    <div className="navbar">
      <nav className="navbar__content">
        <h1 className="navbar__logo">
          <a href="/">
            <img src="../../favicon.png" alt="Pixl logo" />
            <span>Pixl.</span>
          </a>
        </h1>
        <button className="navbar__sign-in-button" type="button">
          <span>Sign in</span>
        </button>
      </nav>
    </div>
  </Fragment>
);

export default NavBar;
