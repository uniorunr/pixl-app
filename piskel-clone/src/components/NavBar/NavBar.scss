.navbar {
  display: flex;
  width: 100%;
  border-bottom: 1px solid #d9d9d9;

  .navbar__content {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    width: 100%;
    margin: 15px 50px;

    .navbar__logo {
      margin: 0;
      flex-grow: 1;

      a {
        color: black;
        text-decoration: none;
        font-size: 2.4rem;
        font-weight: 700;

        img {
          margin-right: 10px;
        }
      }
    }

    .navbar__sign-in-button {
      position: relative;
      top: 0;
      padding: 5px 10px;
      background-color: white;
      border: 2px solid black;
      box-shadow: 2px 2px 0 black;
      cursor: pointer;
      outline: none;

      span {
        font-weight: 700;
      }
    }

    .navbar__sign-in-button:active {
      top: 2px;
      box-shadow: 0 0 0 black;
    }

    .navbar__faq {
      position: relative;
      top: 0;
      margin: 0 20px;
      padding: 5px 10px;
      background-color: white;
      border: 2px solid black;
      box-shadow: 2px 2px 0 black;
      cursor: pointer;
      outline: none;
    }

    .navbar__faq:active {
      top: 2px;
      box-shadow: 0 0 0 black;
    }
  }
}
