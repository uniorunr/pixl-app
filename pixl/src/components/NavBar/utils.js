const updateShortcut = (shortcutObj, code, key, updateFunc, name) => {
  const currShortcutsObj = { ...shortcutObj };
  const shortcuts = Object.keys(currShortcutsObj).map(
    id => currShortcutsObj[id].shortcut,
  );
  if (
    !shortcuts.includes(code)
    && !code.toLowerCase().includes('shift')
    && !code.toLowerCase().includes('alt')
  ) {
    currShortcutsObj[key].shortcut = code;
    updateFunc(currShortcutsObj);
    sessionStorage.setItem(name, JSON.stringify(currShortcutsObj));
  }
};

const keyboardListener = (code, shiftKey, altKey, props) => {
  const {
    modalActive,
    activeTool,
    activeFrameShortcut,
    activeLayerShortcut,
    activeBlock,
    toolsData,
    framesShortcuts,
    layersShortcuts,
    updateCurrToolId,
    updateToolsData,
    updateFramesShortcuts,
    updateLayersShortcuts,
  } = props;
  if (!modalActive && !shiftKey && !altKey) {
    const toolKeys = Object.keys(toolsData);
    const targetTool = toolKeys.find(id => toolsData[id].shortcut === code);
    if (targetTool && !shiftKey && !altKey) {
      updateCurrToolId(targetTool);
    }
  } else if (activeBlock === 'tools' && !shiftKey && !altKey) {
    updateShortcut(
      toolsData,
      code,
      activeTool,
      updateToolsData,
      'toolsData',
    );
  } else if (activeBlock === 'frames' && !altKey && !shiftKey) {
    updateShortcut(
      framesShortcuts,
      code,
      activeFrameShortcut,
      updateFramesShortcuts,
      'framesShortcuts',
    );
  } else if (activeBlock === 'layers' && !altKey && !shiftKey) {
    updateShortcut(
      layersShortcuts,
      code,
      activeLayerShortcut,
      updateLayersShortcuts,
      'layersShortcuts',
    );
  }
};

export default keyboardListener;
