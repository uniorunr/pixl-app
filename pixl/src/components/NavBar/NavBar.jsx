import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import * as actions from '../../actions/actions';
import './NavBar.scss';
import Logo from '../../assets/favicon.png';
import FireBase from '../../firebase/firebase';
import UserInfo from './UserInfo/UserInfo';
import ShortcutsModal from './KeyBoardShortcutsModal/KeyBoardShortcutsModal';

const updateShortcut = (shortcutObj, code, key, updateFunc, name) => {
  const currShortcutsObj = { ...shortcutObj };
  const shortcuts = Object.keys(currShortcutsObj).map(
    id => currShortcutsObj[id].shortcut,
  );
  if (
    !shortcuts.includes(code)
    && !code.toLowerCase().includes('shift')
    && !code.toLowerCase().includes('alt')
  ) {
    currShortcutsObj[key].shortcut = code;
    updateFunc(currShortcutsObj);
    sessionStorage.setItem(name, JSON.stringify(currShortcutsObj));
  }
};

class NavBar extends Component {
  componentDidMount() {
    const {
      layersShortcuts,
      updateCurrToolId,
      updateToolsData,
      updateFramesShortcuts,
      updateLayersShortcuts,
    } = this.props;
    document.addEventListener('keydown', ({ code, shiftKey, altKey }) => {
      const {
        modalActive,
        activeTool,
        activeFrameShortcut,
        activeLayerShortcut,
        activeBlock,
      } = this.props;
      const { toolsData, framesShortcuts } = this.props;
      if (!modalActive && !shiftKey && !altKey) {
        const toolKeys = Object.keys(toolsData);
        const targetTool = toolKeys.find(id => toolsData[id].shortcut === code);
        if (targetTool && !shiftKey && !altKey) {
          updateCurrToolId(targetTool);
        }
      } else if (activeBlock === 'tools' && !shiftKey && !altKey) {
        updateShortcut(
          toolsData,
          code,
          activeTool,
          updateToolsData,
          'toolsData',
        );
      } else if (activeBlock === 'frames' && !altKey && !shiftKey) {
        updateShortcut(
          framesShortcuts,
          code,
          activeFrameShortcut,
          updateFramesShortcuts,
          'framesShortcuts',
        );
      } else if (activeBlock === 'layers' && !altKey && !shiftKey) {
        updateShortcut(
          layersShortcuts,
          code,
          activeLayerShortcut,
          updateLayersShortcuts,
          'layersShortcuts',
        );
      }
    });
  }

  handleSignIn = async () => {
    const { updateLoginStatus } = this.props;
    updateLoginStatus('Signing in...');
    await FireBase.auth();
  };

  handleLogoClick = (e) => {
    e.preventDefault();
    const { changeSection } = this.props;
    changeSection('landing');
    sessionStorage.setItem('section', 'landing');
  };

  render() {
    const {
      userData,
      signInState,
      modalActive,
      updateShortcutsModalState,
    } = this.props;

    return (
      <Fragment>
        <div className="navbar">
          <nav className="navbar__content">
            <h1 role="presentation" className="navbar__logo">
              <a href="/" onClick={this.handleLogoClick}>
                <img src={Logo} alt="Pixl logo" />
                <span>Pixl!</span>
              </a>
            </h1>
            <div className="navbar__info">
              <button
                className="navbar__faq"
                type="button"
                onClick={updateShortcutsModalState}
              >
                <i className="fas fa-question" />
              </button>
              <div className="navbar__tooltip">
                <span>Info</span>
              </div>
            </div>
            {!userData ? (
              <button
                className="navbar__sign-in-button"
                type="button"
                onClick={this.handleSignIn}
              >
                <span>{signInState || 'Sign in'}</span>
              </button>
            ) : (
              <UserInfo src={userData.photoURL} />
            )}
          </nav>
        </div>
        {modalActive ? <ShortcutsModal /> : null}
      </Fragment>
    );
  }
}

NavBar.propTypes = {
  userData: PropTypes.instanceOf(Object),
  signInState: PropTypes.string,
  toolsData: PropTypes.instanceOf(Object).isRequired,
  updateCurrToolId: PropTypes.func.isRequired,
  updateToolsData: PropTypes.func.isRequired,
  framesShortcuts: PropTypes.instanceOf(Object).isRequired,
  layersShortcuts: PropTypes.instanceOf(Object).isRequired,
  updateFramesShortcuts: PropTypes.func.isRequired,
  updateLayersShortcuts: PropTypes.func.isRequired,
  changeSection: PropTypes.func.isRequired,
  updateLoginStatus: PropTypes.func.isRequired,
  modalActive: PropTypes.bool.isRequired,
  activeTool: PropTypes.string.isRequired,
  activeFrameShortcut: PropTypes.string.isRequired,
  activeLayerShortcut: PropTypes.string.isRequired,
  activeBlock: PropTypes.string.isRequired,
  updateShortcutsModalState: PropTypes.func.isRequired,
};

NavBar.defaultProps = {
  userData: null,
  signInState: null,
};

const mapStateToProps = state => ({
  section: state.section,
  signInState: state.firebase.signInState,
  userData: state.firebase.userData,
  toolsData: state.tools.toolsData,
  framesShortcuts: state.frames.framesShortcuts,
  layersShortcuts: state.layers.layersShortcuts,
  modalActive: state.shortcuts.modalActive,
  activeTool: state.shortcuts.activeTool,
  activeBlock: state.shortcuts.activeBlock,
  activeFrameShortcut: state.shortcuts.activeFrameShortcut,
  activeLayerShortcut: state.shortcuts.activeLayerShortcut,
});

const mapDispatchToProps = (dispatch) => {
  const {
    changeSection,
    updateLoginStatus,
    updateCurrToolId,
    updateToolsData,
    updateFramesShortcuts,
    updateLayersShortcuts,
    updateShortcutsModalState,
  } = bindActionCreators(actions, dispatch);
  return {
    changeSection: (section) => {
      changeSection(section);
    },
    updateLoginStatus: (status) => {
      updateLoginStatus(status);
    },
    updateCurrToolId: (id) => {
      updateCurrToolId(id);
    },
    updateToolsData: (data) => {
      updateToolsData(data);
    },
    updateFramesShortcuts: (data) => {
      updateFramesShortcuts(data);
    },
    updateLayersShortcuts: (data) => {
      updateLayersShortcuts(data);
    },
    updateShortcutsModalState,
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(NavBar);
