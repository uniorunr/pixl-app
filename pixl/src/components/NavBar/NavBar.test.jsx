import React from 'react';
import renderer from 'react-test-renderer';
import NavBar from './NavBar';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <NavBar
        framesShortcuts={{}}
        layersShortcuts={{}}
        toggleSection={() => {}}
        toolsData={{}}
        updateCurrentTool={() => {}}
        updateFrameShortcuts={() => {}}
        updateLayersShortcuts={() => {}}
        updateToolsData={() => {}}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
