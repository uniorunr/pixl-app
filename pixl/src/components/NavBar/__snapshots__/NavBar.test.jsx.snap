// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="navbar"
>
  <nav
    className="navbar__content"
  >
    <h1
      className="navbar__logo"
      onClick={[Function]}
      role="presentation"
    >
      <a
        href="/"
      >
        <img
          alt="Pixl logo"
          src="test-file-stub"
        />
        <span>
          Pixl!
        </span>
      </a>
    </h1>
    <div
      className="navbar__info"
    >
      <button
        className="navbar__faq"
        onClick={[Function]}
        type="button"
      >
        <i
          className="fas fa-question"
        />
      </button>
      <div
        className="navbar__tooltip"
      >
        <span>
          Info
        </span>
      </div>
    </div>
    <button
      className="navbar__sign-in-button"
      onClick={[Function]}
      type="button"
    >
      <span>
        Sign in
      </span>
    </button>
  </nav>
</div>
`;
